ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/worldstate/MarkSweepPruner.java

```diff
@@ -0,0 +1,177 @@
+import tech.pegasys.pantheon.ethereum.trie.StoredMerklePatriciaTrie;
+import tech.pegasys.pantheon.metrics.Counter;
+import tech.pegasys.pantheon.metrics.MetricsSystem;
+import tech.pegasys.pantheon.metrics.PantheonMetricCategory;
+import tech.pegasys.pantheon.services.kvstore.KeyValueStorage;
+import tech.pegasys.pantheon.services.kvstore.KeyValueStorage.Transaction;
+import tech.pegasys.pantheon.util.bytes.Bytes32;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class MarkSweepPruner {
+  private static final Logger LOG = LogManager.getLogger();
+  private static final BytesValue IN_USE = BytesValue.of(1);
+  private static final int MARKS_PER_TRANSACTION = 1000;
+  private final WorldStateStorage worldStateStorage;
+  private final KeyValueStorage markStorage;
+  private final Counter markedNodesCounter;
+  private final Counter markOperationCounter;
+  private final Counter sweepOperationCounter;
+  private final Counter sweptNodesCounter;
+  private volatile long nodeAddedListenerId;
+  private final ReentrantLock markLock = new ReentrantLock(true);
+  private final Set<BytesValue> pendingMarks = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+  public MarkSweepPruner(
+      final WorldStateStorage worldStateStorage,
+      final KeyValueStorage markStorage,
+      final MetricsSystem metricsSystem) {
+    this.worldStateStorage = worldStateStorage;
+    this.markStorage = markStorage;
+
+    markedNodesCounter =
+        metricsSystem.createCounter(
+            PantheonMetricCategory.PRUNER,
+            "marked_nodes_total",
+            "Total number of nodes marked as in use");
+    markOperationCounter =
+        metricsSystem.createCounter(
+            PantheonMetricCategory.PRUNER,
+            "mark_operations_total",
+            "Total number of mark operations performed");
+
+    sweptNodesCounter =
+        metricsSystem.createCounter(
+            PantheonMetricCategory.PRUNER,
+            "swept_nodes_total",
+            "Total number of unused nodes removed");
+    sweepOperationCounter =
+        metricsSystem.createCounter(
+            PantheonMetricCategory.PRUNER,
+            "sweep_operations_total",
+            "Total number of sweep operations performed");
+  }
+
+  public void prepare() {
+    worldStateStorage.removeNodeAddedListener(nodeAddedListenerId); // Just in case.
+    nodeAddedListenerId = worldStateStorage.addNodeAddedListener(this::markNewNodes);
+  }
+
+  public void mark(final Hash rootHash) {
+    markOperationCounter.inc();
+    markStorage.clear();
+    createStateTrie(rootHash)
+        .visitAll(
+            node -> {
+              if (Thread.interrupted()) {
```

Comment:
Is checking the interrupt the best way to stop an in-process mark?  Should we consider a shared boolean instead?  Also, do we have a test case covering this abort scenario?  Is it one we expect often?  Or just during shutdown?

Reply:
Since we're not looking to ever abort a mark ourselves, my understanding is the only thing we need to make sure to check is if we're interrupted. Does that make sense?

Reply:
That makes sense, we are only looking for JVM shutdown.  May be worth a code comment in that case.